// angular-recaptcha3/esm2015/recaptcha.service.js
import { Injectable, NgZone, Optional, SkipSelf } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
/*
 * Common service shared by all reCaptcha component instances
 * through dependency injection.
 * This service has the task of loading the reCaptcha API once for all.
 * Only the first instance of the component creates the service, subsequent
 * components will use the existing instance.
 *
 * As the language is passed to the <script>, the first component
 * determines the language of all subsequent components. This is a limitation
 * of the present Google API.
 */
export class ReCaptchaService {
    constructor(zone) {
        this.scriptLoaded = false;
        this.readySubject = new BehaviorSubject(false);
        /* the callback needs to exist before the API is loaded */
        if (typeof window !== 'undefined') {
            window['reCaptchaOnloadCallback'] = (() => {
                zone.run(this.onloadCallback.bind(this));
            });
        }
    }
    getReady(language) {
        if (!this.scriptLoaded) {
            this.scriptLoaded = true;
            let doc = document.body;
            let script = document.createElement('script');
            script.innerHTML = '';
            script.src = 'https://www.google.com/recaptcha/api.js?onload=reCaptchaOnloadCallback&render=explicit' +
                (language ? '&hl=' + language : '');
            script.async = true;
            script.defer = true;
            doc.appendChild(script);
        }
        return this.readySubject.asObservable();
    }
    execute(options = undefined) {
        const captcha = window.grecaptcha;
        const captchaElement = document.querySelector('.g-recaptcha.invisible');
        return captcha.execute(captchaElement.id, options);
    }
    onloadCallback() {
        this.readySubject.next(true);
    }
}
ReCaptchaService.decorators = [
    { type: Injectable }
];
ReCaptchaService.ctorParameters = () => [
    { type: NgZone }
];
/* singleton pattern taken from https://github.com/angular/angular/issues/13854 */
export function RECAPTCHA_SERVICE_PROVIDER_FACTORY(ngZone, parentDispatcher) {
    return parentDispatcher || new ReCaptchaService(ngZone);
}
export const RECAPTCHA_SERVICE_PROVIDER = {
    provide: ReCaptchaService,
    deps: [NgZone, [new Optional(), new SkipSelf(), ReCaptchaService]],
    useFactory: RECAPTCHA_SERVICE_PROVIDER_FACTORY
};
